### PR TITLE
Support change the LB selection mode on the existing serivces

### LINKED ISSUE #416: Mode selection annotation `service.beta.kubernetes.io/azure-load-balancer-mode` only works when the service is newly created
**What happened**:
Mode selection annotation `service.beta.kubernetes.io/azure-load-balancer-mode` only works when the service is newly created. It doesn't work if we modify this annotation after the service has been created.

**What you expected to happen**:
The annotation works every time.

**How to reproduce it**:
Create a service and modify the annotation `service.beta.kubernetes.io/azure-load-balancer-mode`.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by nilo19**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
Support change the LB selection mode annotation on the existing services.

**Which issue(s) this PR fixes**:

Fixes: https://github.com/kubernetes-sigs/cloud-provider-azure/issues/416
Ref: https://github.com/kubernetes-sigs/cloud-provider-azure/issues/427

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Support change the LB selection mode on the existing serivces
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```
/sig cloud-provider
/area provider/azure
/priority important-soon
/triage accepted

### CODE REVIEW

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go [hunk 1]
```diff
@@ -2504,25 +2597,15 @@ func subnet(service *v1.Service) *string {
 // getServiceLoadBalancerMode parses the mode value.
 // if the value is __auto__ it returns isAuto = TRUE.
 // if anything else it returns the unique VM set names after trimming spaces.
-func getServiceLoadBalancerMode(service *v1.Service) (hasMode bool, isAuto bool, vmSetNames []string) {
+func (az *Cloud) getServiceLoadBalancerMode(service *v1.Service) (hasMode bool, isAuto bool, vmSetName string) {
```
Comment (feiskyer) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:2600]: would `AS1,AS2"` be still supported?

**Review by feiskyer — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go [hunk 2]
```diff
@@ -351,6 +352,120 @@ func (az *Cloud) cleanBackendpoolForPrimarySLB(primarySLB *network.LoadBalancer,
...
+	lbName := strings.TrimSuffix(currLBName, InternalLoadBalancerNameSuffix)
+	if strings.EqualFold(lbName, vmSetName) {
+		return false
+	}
+	if strings.EqualFold(vmSetName, az.VMSet.GetPrimaryVMSetName()) && strings.EqualFold(clusterName, lbName) {
+		return false
+	}
+	// if the VMSS/VMAS of the current LB is different from the mode, change the LB
+	// to another one
+	klog.V(2).Infof("shouldChangeLoadBalancer(%s, %s, %s): change the LB to another one", service.Name, currLBName, clusterName)
```
[the remark this answers is not in the document]
Comment (nilo19) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:374]: the new lb name is not defined here

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go (hunk 1 above)
Comment (nilo19) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:2600]: no, it would failed in `GetVMSetNames`

**Review by feiskyer — APPROVED** [review]
/lgtm
/retest

**Review by andrewsykim — COMMENTED** [review]
Is this a really bug or a feature? It sounds like _not_ supporting the annotation for existing Services was an intentional behavior from the start.

**Comment by andrewsykim** [comment]
I'm putting a temporary hold on this to discuss this further. As previously mentioned, we are freezing features to legacy-cloud-providers starting in v1.21 so I would like to understand if this is really a bug fix or feature. 

/hold

**Comment by nilo19** [comment]
@andrewsykim this is not an intentional behavior at the beginning. This is a bug. We found it during the code freeze.

**Comment by feiskyer** [comment]
yes, this is a bug fix for the annotation that exists for a long time.

**Comment by nilo19** [comment]
If we cannot merge the bugfix, I would close it and do it in our fork in out-of-tree repo.

**Comment by andrewsykim** [comment]
What's the implication of changing a load balancer's "mode"? Does that provision a new load balancer instance or change some configuration on an existing one?

**Comment by nilo19** [comment]
> What's the implication of changing a load balancer's "mode"? Does that provision a new load balancer instance or change some configuration on an existing one?

It would re-associate the lb of the service to a new one. Actually, there is one lb per agent pool, and changing the mode means we could change the service's lb instance.

**Pull request closed without merging**